src/main/java/dev/waterdog/waterdogpe/ProxyServer.java

```diff
@@ -190,9 +191,22 @@ private void boot() {
         }
 
         this.bedrockServer = new BedrockServer(bindAddress, Runtime.getRuntime().availableProcessors(), this.bossEventLoopGroup, this.workerEventLoopGroup, false);
-        this.bedrockServer.setHandler(new ProxyListener(this));
+        this.bedrockServer.setHandler(new ProxyListener(this, this.queryHandler, bindAddress));
         this.bedrockServer.bind().join();
 
+        for (Integer port : this.getConfiguration().getAdditionalPorts()) {
+            logger.info("Starting additional port " + port);
+            InetSocketAddress additionalBind = new InetSocketAddress(bindAddress.getAddress(), port);
+
+            BedrockServer newServer = new BedrockServer(additionalBind, Runtime.getRuntime().availableProcessors(), this.bossEventLoopGroup, this.workerEventLoopGroup, false);
+            newServer.setHandler(new ProxyListener(this, new QueryHandler(this, additionalBind), additionalBind));
```

Comment:
Any reason why we need more instances of  the `QueryHandler`? It would be better to share one common handler instance over the `BedrockServer`s. You can get the current server address inside of `QueryHandler#onQuery` from `ctx.localAddress()` or just pass it as an extra argument when calling the method.

Reply:
Fixed in [9148026](https://github.com/WaterdogPE/WaterdogPE/pull/216/commits/914802629568357882879785f23e8dac3a7c2599)